src/epd_driver/epd_driver.c

```diff
@@ -40,13 +41,73 @@ void epd_draw_vline(int x, int y, int length, uint8_t color,
   }
 }
 
+EpdRect _rotated_area(uint16_t x, uint16_t y, uint16_t w, uint16_t h) {
+  // If partial update uses full screen do not rotate anything
+  if (x != 0 && y != 0 && EPD_WIDTH != w && EPD_HEIGHT != h) {
+    switch (epd_get_rotation())
+    {
+      // 0 landscape: Leave it as is
+      // 1 90 ° right portrait
+      case 1:
+        _swap_int(x, y);
+        _swap_int(w, h);
+        x = EPD_WIDTH - x - w;
+        break;
+      
+      case 2:
+        // 3 180° landscape
+        x = EPD_WIDTH - x - w;
+        y = EPD_HEIGHT - y - h;
+        break;
+
+      case 3:
+        // 3 270 ° portrait
+        _swap_int(x, y);
+        _swap_int(w, h);
+        y = EPD_HEIGHT - y - h;
+        break;
+    }
+  }
+  //printf("AFT x %d y %d w %d h %d\n\n",x,y,w,h); // Debug: Remove after tests
+  EpdRect rotated =  {
+    x, y, w, h
+  };
+  return rotated;
+}
+
+Coord_xy _rotate(uint16_t x, uint16_t y) {
+    switch (display_rotation) {
+        case 1:
+            _swap_int(x, y);
+            x = EPD_WIDTH - x - 1;
+        break;
+        case 2:
+            x = EPD_WIDTH - x - 1;
+            y = EPD_HEIGHT - y - 1;
+        break;
+        case 3:
+            _swap_int(x, y);
+            y = EPD_HEIGHT - y - 1;
+        break;
+    }
+    Coord_xy coord = {
+        x, y
+    };
+    return coord;
+}
+
 void epd_draw_pixel(int x, int y, uint8_t color, uint8_t *framebuffer) {
-  if (x < 0 || x >= EPD_WIDTH) {
+  if (x < 0 || x >= epd_rotated_display_width()) {
     return;
   }
-  if (y < 0 || y >= EPD_HEIGHT) {
+  if (y < 0 || y >= epd_rotated_display_height()) {
```

Comment:
If I'm not mistaken, these checks should be moved to after the rotation and then check for EPD_WIDTH / HEIGHT as is. This might be equivalent, but to me more confusing since these checks are meant to guard for the actual buffer bounds.

Reply:
This is the only point that I'm not sure how to solve. Can you specify better how would you do it?